---
title: "Shine a Light on MAGIC Data"
author: "Brian S Yandell, <http://www.stat.wisc.edu/~yandell>"
date: "Jackson Labs, 18 October 2016"
output:
  beamer_presentation:
    includes:
      in_header: page_number.tex
  ioslides_presentation:
    widescreen: no
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## shine a light on MAGIC data

Shiny tool to study DO and other MAGIC crosses

- What is MAGIC?
- Allele vs SNP Scans
- Additive Model is Quick & Easy
- Gene/SNP Action
- Shiny under the Hood

## what is MAGIC?

- Multiparent Advanced Generation Inter-Cross (MAGIC) 
    + experimental populations with >2 segregating alleles
- advanced generations yield many meiotic events
    + typically low linkage disequilibrium
    + capable of fine mapping in one pass
- de Koning DJ, McIntyre LM ([G3 2014](http://dx.doi.org/10.1534/g3.114.013680))
    + animals: mouse, Drosophila
    + plants: maize, wheat, rice, sorghum, Arabidopsis, Pigeonpea
    + mapping populations: AIL, CC, HS, DO, DSPR

## Attie/Jax DO population

- 8 CC founder strains (generation 19-21)
- 400 mice in 4 waves
- multiple traits measured
    * clinical traits (insulin secretion)
    * 150K SNPs, 30K RNA-Seq
    * proteomic, metabolomic, lipidomic

![Valdar et al. 2006 doi:10.1534/genetics.104.039313](images/valdar2006.gif)
    
----

![http://compgen.unc.edu](images/ccmice.png)

## allele vs SNP scans

- allele-based genome scan: LOD maps
    + continuous curve across loci
    + interval mapping for missing data
    + model effect of founder alleles
- DO founder alleles: A,B,C,D,E,F,G,H
- response ~ sum of effects of alleles
- additive model
    + $y$ ~ $a_1 + a_2$ (1st & 2nd allele)
    + test if all a's are the same
    + 8 unknown parameters

## simple story on chr 1

![chr 1 scan](images/scan_1.pdf)
    
----

![chr 1 zoom scan & effects](images/scan_1_effect.pdf)

## SNP-based genome scan

- SNP-based genome scan: GWAS Manhattan plots
    + discrete tests of SNPs or other features
    + typically 2 SNP alleles (1 is reference)
    + model effect of number of non-ref SNP copies 
- SNP recorded as pair of DNA base pairs (A,C,G,T)
    + SNPs typically have two values (G/T)
    + individual has genotype GG, GT or TT
- simplifed to number of copies of non-reference allele
    + $s$ = 0,1,2
    + DO reference is B = B6
- additive model:
    + $y$ ~ $a + bs$ (2 unknowns)
    + test slope: $b=0$?

----

![SNP scan & top pattern](images/snp_assoc_1_pattern.pdf)

----

![LOD & allele contrast scan](images/snp_allele_1_pattern.pdf)

## additive model is quick & easy

- good news
    + fewer parameters to understand
    + easy to build tools
    + fast to run
    + relate LOD to allele effects directly
    + likely good model for mRNA expression
- bad news
    + could miss important gene action information
    + could miss loci (false negatives)
    + could detect false positives
    + LOD and allele effects may conflict due to dominance

## full model (with dominance)

- DO founder alleles: A,B,C,D,E,F,G,H
- response ~ effect of pair of alleles
- full model for allele-based scan
    + $y$ ~ $a_1 + a_2 + d_{12}$ (additive & dominance effects)
    + $y$ ~ $\texttt{mean}(A_1,A_2)$ (alleles $A_1$, $A_2$)
    + test if all means are the same (all $a$s equal, all $d$s zero)
    + 36 unknown parameters
- full model for SNP-based scan
    + $y$ ~ $\texttt{mean}(s)$
    + test if all means are the same
    + 3 unknowns

## gene/SNP action

- study additive & dominance of single trait
- compare co-mapping traits -- same gene action?
- extend discrete SNP-based scan to continuous
    + scan of particular allele contrasts
    + LOD and allele contrast scans
    
```{r echo=FALSE, fig.height=3}
data <- data.frame(x=rep(0:2,4),y=c(0:2,0,1.5,2,0,2,2,0,0,2),
                   z=rep(c("additive","general","dominant","recessive"), each = 3))
library(ggplot2)
ggplot(data, aes(x,y)) +
  geom_point(size=4) +
  geom_line(size=2) +
  facet_grid(~z) +
  labs(x="",y="phenotype mean") + 
  theme(axis.text.y = element_blank()) +
  scale_x_continuous(breaks=0:2)
```

## complicated story on chr 3

![chr 3 scan](images/scan_3.pdf)
    
----

![chr 3 zoom scan & effects](images/scan_3_effect.pdf)

----

![SNP scan & top pattern](images/snp_assoc_3_pattern.pdf)

## additive & dominance together

![3-level SNP scan & top pattern](images/snp_assoc_3_pattern_add+dom.pdf)

## general SNP contrasts

- identify 3-level SNP with strong LOD
- interpret as contrast involving 36 allele pairs
    + example AB:CDEFGH
- divide subjects into 3 groups at each locus
    + AA,AB,BB allele pairs (3)
    + CC,CD,...,GH,HH allele pairs (21)
    + rest allele from AB and allele from CDEFGH (12)
- scan region for this set of contrasts
    + LOD scan with 2 df of AB:het:CDEFGH
    + allele group scan

## additive BFGH:ACDE 

![](images/snp_allele_3_BEFGH:ACD_additive.pdf)

## B=B6 recessive, D=NOD dominant

![](images/snp_allele_3_ABCEFGH:D_recessive.pdf)

## B=B6 dominant, G=PWK recessive

![unstable: few double recessive GG](images/snp_allele_3_ABCDEFH:G_dominant.pdf)


## shiny under the hood

- work flow
- tools and resources
- challenges remaining

![[lambda architecture](http://lambda-architecture.net/) (Nathan Marz, Twitter)](images/la-overview_small.png)

## lambda architecture for doqtl2

- batch layer (DOQTL)
    + genotypes
    + phenotype LOD scans across genome
    + phenotype LOD & allele scans for chromosome
    + phenotype SNPs & genes in peak region
- serving layer
    + [web serving layer](https://www.biostat.wisc.edu/~kbroman/attieDO/lodpeaks.html)
- speed layer
    + [Shiny \"speed\" layer](http://www.garyzone.com/shiny/doqtl2/doqtl2/)
    + phenotype LOD & allele scan for chromosome
    + detailed look at peak region

## tools & resources

- Original batch pipeline: [DOQTL](http://dx.doi.org/10.1534/g3.114.013748) (Dan Gatti et al. 2014 G3)
- [R/qtl2](http://kbroman.org/qtl2/assets/vignettes/user_guide.html) (Karl Broman) (Karl Broman et al. in progress)
    + qtl2geno, qtl2scan, qtl2plot
    + [web site](https://github.com/kbroman/qtl2/tree/gh-pages)
- new R packages
    + [R/qtl2 package suite](http://kbroman.org/qtl2/assets/vignettes/user_guide.html) (Karl Broman)
    + [R/doqtl2](https://bitbucket.org/byandell/doqtl2) package
    + [R/qtl2shiny](https://bitbucket.org/byandell/qtl2shiny) package
- Derived Data in various file forms
    + SQLite for mouse SNPs and gene features
    + RDS R data objects for genotypes
    + CSV comma separated variable for phenotypes

## challenges remaining

- Shiny user interface from Rstudio
    + translates from R to HTML via javascript
    + run under Rstudio or via Shiny server
    + R/qtl2shiny has 20+ shiny modules
    + translate to Python?
- analysis & computation
    + calibration of significance thresholds
    + dimension reduction (filtering) for massive phenotypes
    + incorporation of causal network tools
- operation & connectivity
    + connecting to other omic resources
    + fast management of raw, derived and intermediate data
- visualizations
    + scalable & interactive (D3) visual displays
    + automated report generation
    
## data storage issues

- raw data
- derived data
    + 3Gb gene, SNP & SVS features (mouse)
    + 2Gb haplotype probabilities
    + 10Gb diplotype probabilities
    + 0.2Mb clinical phenotypes
    + 1-2Gb molecular phenotypes
    + spatial/image phenotypes?
- intermediate data
    + tables & summaries
    + plot data (or saved plots)
    + on-the-fly vs pre-stored vs as-needed
- portability issues
    + CSV vs RDS vs SQL vs ...
    + common pool vs on-site
    
## software issues

- R analysis & visualization
    + QTL tools: qtl2 squite: geno, scan, plot
    + discovery: doqtl2, qtl2shiny
    + interactive: shiny, shinydashboard
    + wrangle data: dplyr, tidyr, readr, stringr
    + graphics: ggplot2, grid
- phenotype & genotype pipelines
    + variety of languages and formats
    + in domain of biologists (or chemists or ...)
- high volume pipelines
    + hadoop/map-reduce technology
    + high throughput phenotypes
    + data resampling for thresholds
- scaling up to muli-user interactive system
    