---
title: "qtl2shiny Developer Guide"
author: "Brian S. Yandell"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = ""
)
```

This document is a Developer Guide for the [R/qtl2shiny](https://github.com/byandell/qtl2shiny) package, with goals to explain:

- data structures and folder organization
- package workflow and shiny tool components

Those wishing to _use_ the shiny interface with their own data will want to read about the data structures, and contact me. Those wishing to _extend_ capabilities of the package, or understand its inner working, with be interested in the package organization.

## Data Structures and Folder Organization

The interface assumes the Shiny app file [app.R](https://github.com/byandell/qtl2shiny/blob/master/inst/qtl2shinyApp/app.R) is in a directory with other support documents and the data. 

The CSV file [`projects.csv`](https://github.com/byandell/qtl2shiny/blob/master/inst/qtl2shinyApp/projects.csv), in that same directory, contains project information. This file has the following content:

```
project,taxa,directory
Recla,CCmouse,qtl2shinyData
```

The column names are fixed. The `project` is the project name (`Recla`), `taxa` has the taxa name (`CCmouse`), and `directory` is the relative or absolute address of the directory where data are kept (`qtl2shinyData`, which is relative here). In this case, the basic structure of the app is:

[`app.R`](https://github.com/byandell/qtl2shiny/blob/master/inst/qtl2shinyApp/app.R)  
[`about.md`](https://github.com/byandell/qtl2shiny/blob/master/inst/qtl2shinyApp/about.md)  
[`about-extended.md`](https://github.com/byandell/qtl2shiny/blob/master/inst/qtl2shinyApp/about-extended.md)  
[`projects.csv`](https://github.com/byandell/qtl2shiny/blob/master/inst/qtl2shinyApp/projects.csv)  
`qtl2shinyData/`  
`qtl2shinyData/CCmouse/`  
`qtl2shinyData/CCmouse/Recla/`

The `about.md` and `about-extended.md` files are place-holders for now for online documentation. All data are now stored under the directory (`qtl2shinyData`) using relative addresses. However, it is possible for data for a `taxa` and `project` to be stored in another place, in which case an absolute address would be supplied in the `projects.csv` file.

Detailed information about the `taxa` and `project` can be found with the [CCmouse.Rmd](https://github.com/byandell/qtl2shiny/blob/master/inst/doc/CCmouse.Rmd) and [Recla.Rmd](https://github.com/byandell/qtl2shiny/blob/master/inst/doc/Recla.Rmd) examples, respectively.

### Taxa information

The `taxa` sub-directory (`CCmouse`) contains the following files. See <https://github.com/rqtl/qtl2db>
for explanation of the SQL databases and query routines.

file | contents
-----|---------
`allele_info.rds` | information on alleles for taxa
`cc-variants.sqlite` | CC structural variants in SQL database
`mouse_genes_mgi.sqlite` | mouse genes curated by MGI in SQL database
`query_genes.rds` | query routine for genes
`query_variants.rds` | query routines for structural variants

### Project Information

The `project` sub-sub-directory (`Recla`) has the data specific to the project. An example of how this is created can be found in [Recla.Rmd](https://github.com/byandell/qtl2shiny/blob/master/inst/doc/Recla.Rmd).
Most of the objects are stored as [RDS](https://www.rdocumentation.org/packages/base/versions/3.4.3/topics/readRDS).
The genotype probabilities are converted from `calc_genoprob` to `feather_genoprob` format using [R/qtl2feather](https://github.com/byandell/qtl2feather).
Query routines are somewhat analogous to those used for `taxa` above.

Note that [R/qtl2](http://www.rqtl.org/qtl2) wants data files to be consistent in terms of individual identifiers. The `covar.rds` and `pheno_data.rds` files have identifiers as row names, which are also used in the genotype probabilities.

file | contents
-----|------------------------------------------------
`genoprob/` | directory with genotype probabilities created with [R/qtl2feather](https://github.com/byandell/qtl2feather)
`kinship.rds` | kinship object using LOCO (see [R/qtl2](http://kbroman.org/qtl2/assets/vignettes/user_guide.html#calculating_a_kinship_matrix))
`pmap.rds` | physical map
`analyses.rds` | data frame of information about analyses (one row per phenotype)
`covar.rds` | data frame of covariates (row = individual, column = covariate)
`pheno_data.rds` | matrix of phenotype data (row = individual, column = phenotype)
`peaks.rds` | data frame of peak information (one row per peak)
`hotspot.rds` | hotspot object with counts of peaks by position and phenotype set
`query_mrna.rds` | query of mRNA data
`query_probs.rds` | query of genotype probabilities

Likely changes:

- pheno_data.rds may become a feather database
- analyses.rds will be redesigned

Some projects have an `RNAseq` folder. This, and accompanying routines for mRNA now in [R/DOread](https://github.com/byandell/DOread) and other packages, will likely be redesigned. Currently, code is set up for only one type of mRNA data.

## Data files


## Query files

The query files are designed to encapsulate specific information about where data are stored and how to retrieve them, so that different projects might use different strategies. For instance, some projects might use CSV or RDS files, while others might use SQL or feather databases. There are four query objects, for genes, variants, mrna and probs.

Note that if you use a similar format as [CCmouse.Rmd](https://github.com/byandell/qtl2shiny/blob/master/inst/doc/CCmouse.Rmd) and [Recla.Rmd](https://github.com/byandell/qtl2shiny/blob/master/inst/doc/Recla.Rmd) examples
to create and store the queries, you may have to quit RStudio and restart it to have it properly pick up the `project` or `taxa` directory.

The following two `create` functions are in [R/qtl2](http://www.rqtl.org/qtl2). They create functions (`query_genes` and `query_variants`) that contain, in their local environment, the location of the respective SQL database. Saving these as RDS preserves this environment information for later reuse.

```{r eval=FALSE}
query_genes <- 
  qtl2::create_gene_query_func("CCmouse/mouse_genes.sqlite")
saveRDS(query_genes, "query_genes.rds")
```

```{r eval=FALSE}
query_variants <- 
  qtl2::create_variant_query_func("CCmouse/cc_variants.sqlite"))
saveRDS(query_variants, "query_variants.rds")
```

This allows calls of the following form to be embedded in code, without reference to where the database is, or even what type of database it might be. Here we look on chromosome `"1"` between `39` and `40` Mbp.

```{r eval=FALSE}
genes <- query_genes("1", 39, 40)
variants <- query_variants("1", 39, 40)
```

The above two routines would generally be used across all projects involving the CC mouse. The following two routines are project specific, refering to genotype probabilities and mRNA data. They do not directly identify the data file, as that is currently hard-coded into the [R/DOread](https://github.com/byandell/DOread) package. These will change substantially in the near future.

```{r eval=FALSE}
query_probs <- DOread::create_probs_query_func_do("CCmouse/Recla")
saveRDS(query_probs, "query_probs.rds")
```

```{r eval=FALSE}
query_mrna <- DOread::create_probs_query_func_do("CCmouse/Recla")
saveRDS(query_probs, "query_mrna.rds")
```

Again, these enable calls to access data in a region. I include additional arguments that are used for these routines. The call, respectively, `DOread::read_probs` and `DOread::read_mrna`.

```{r eval=FALSE}
probs <- query_probs("1", 39, 40, allele = TRUE, method = "feather")
mrna <- query_mrna("1", 39, 40, local = TRUE, qtl = FALSE)
```

## Package Workflow and Shiny Tool Components

The `qtl2shiny` interface can be invoked within R using [runApp](https://shiny.rstudio.com/reference/shiny/latest/runApp.html) on the file
[app.R](https://github.com/byandell/qtl2shiny/blob/master/inst/qtl2shinyApp/app.R). There is a routine [qtl2shinyApp](https://github.com/byandell/qtl2shiny/blob/master/R/qtl2shinyApp.R) in the package to do this as well:

```{r}
qtl2shiny::qtl2shinyApp
```

However, one first should set up the data and query files as described above.

The shiny server consists of the server app (`server()` and `ui()` in file `app.R`) and 27 shiny modules. The modules form a directed acyclic graph with `Main` being the root node.

```{r echo=FALSE, child="../inst/scripts/qtl2dag.Rmd"}

```


![](../inst/scripts/net.jpg)

The `Main` module calls the `Setup`, `Haplo` and `Diplo` modules, corresponding to the three dashboard menu items. Each of these modules in turn calls other modules. The modules were each designed to do one task, and be kept small, ideally about 100 lines of code (including documentation). Here is a list of modules:

modeul | lines | task            | output
------|--------|-----------------|-------
[`Main`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyMain.R) | 129 | call Setup, Haplo, Diplo |
[`Setup`](https://github.com/byandell/qtl2shiny/blob/master/R/shinySetup.R) | 209 | set up project |
[`Haplo`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyHaplo.R) | 103 | Haplotype Scans |
[`Diplo`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyDiplo.R) | 107 | SNP/Gene Action |
 | | |
[`Project`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyProject.R) | 47 | pick project | 
[`Peaks`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyPeaks.R) | 175 | select chr, peak and window width |
[`Phenos`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyPhenos.R) | 88 | phenotype information | table
[`Hotspot`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyHotspot.R) | 190 | hotspot count by location | plot, table
[`PhenoPlot`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyPhenoPlot.R) | 66 | plot phenotypes | plot
 | | |
[`Probs`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyProbs.R) | 93 | allele genotype probabilities |
[`ScanCoef`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyScanCoef.R) | 210 | allele genome LOD and effect scans | plot, table
[`Mediate`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyMediate.R) | 306 | mediation by other phenotypes | plot, table
[`Scatter`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyScatter.R) | 149 | scatter plot for mediator | plot
 | | |
[`PairProbs`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyPairProbs.R) | 93 | allele pair genotype probabilities |
[`Pattern`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyPattern.R) | 247 | LOD and imputed SNP effects by pattern | plot, table
[`Allele`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyAllele.R) | 126 | allele and allele pair means | plot, tablt
 | | |
[`SNPSetup`](https://github.com/byandell/qtl2shiny/blob/master/R/shinySNPSetup.R) | 225 | set up SNP processng |
[`SNPPattern`](https://github.com/byandell/qtl2shiny/blob/master/R/shinySNPPattern.R) | 276 | SNP allele patterns | plot, table
[`SNPFeature`](https://github.com/byandell/qtl2shiny/blob/master/R/shinySNPFeature.R) | 113 | top SNP features | plot, table
[`SNPProbs`](https://github.com/byandell/qtl2shiny/blob/master/R/shinySNPProbs.R) | 93 | SNP genotype probabilities |
[`SNPGene`](https://github.com/byandell/qtl2shiny/blob/master/R/shinySNPGene.R) | 102 | SNP & Gene Choice | table
[`SNPPlot`](https://github.com/byandell/qtl2shiny/blob/master/R/shinySNPPlot.R) | 64 | plot SNP association | plot
[`SNPSum`](https://github.com/byandell/qtl2shiny/blob/master/R/shinySNPSum.R) | 143 | create table of top SNPs | table
[`GeneRegion`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyGeneRegion.R) | 115 | genes and SNPs in region | plot, table
[`GeneExon`](https://github.com/byandell/qtl2shiny/blob/master/R/shinyGeneExon.R) | 162 | exons for individual genes | plot, table
