---
title: "R/qtl2shiny Screen Shots"
author: "Brian S. Yandell"
date: "`r format(Sys.Date(), '%d %b %Y')`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

This shows a series of screen shots, giving examples of `R/qtl2shiny` in action. For more details on use of the package, see the 
[R/qtl2shiny package](https://github.com/byandell/qtl2shiny),
[R/qtl2shiny User Guide](https://github.com/byandell/qtl2shiny/blob/master/vignettes/UserGuide.Rmd) and [R/qtl2shiny Developer Guide](https://github.com/byandell/qtl2shiny/blob/master/vignettes/DeveloperGuide.Rmd). This document is created by [inst/figs/index.Rmd](https://github.com/byandell/qtl2shiny/blob/master/inst/figs/index.Rmd) and is saved as HTML at [~yandell/software/qtl2shiny](http://www.stat.wisc.edu/~yandell/qtl2shiny).


Initial screen shot. The `chr` and `pos` are preset by largest hotspot (region with most LOD peaks above threshold) with default window of 1Mb. In pink boxes in side dashboard and side panel, notice project name and `chr` and `pos`.

![](version.png)

## Phenotypes and Region

User can choose phenotypes, which are ordered by largest peak in selected window. Once a phenotype is selected, it is removed from the choice list.

![](choose_phenotype.png)

![](region_phenotype.png)

Switching top button to `Phenotypes` gives choices of displays for phenotypes. The initial one is `LOD Peaks`.

![](phenotypes.png)

Selecting `Covariates` displays aspects of analysis, including names of covariates.

![](pheno_covariates.png)

Analysis is done using the data `transf`ormation. You can see a histogram of the data using `Trans Data`

![](pheno_trans.png)

or see the raw data using `Raw Data`.

![](pheno_raw.png)

Now return to `Region` and look at hotspots across whole genome by checking the `plot?` box in the `Hotspot Info` area of the side panel. Uses window width of 1Mb to count peaks above LOD 5.5. Both are tunable.

![](hotspots.png)

Look only for chromosome 11. Notice how summary table changes to phenotypes on chromosome from counts by chromosome.

![](hotspots_chr_11.png)

Switching the focus of `Hotspot Info` will change the focal `chr` and `pos`. This in turn changes the table at upper right to show the phenotypes with strongest LOD.

![](hotspots_chr_1.png)

As before, you can choose phenotypes driven by a chromosome hotspot. Notice that with the hotspot infor now focused on one chr, the second table on the right is reduced. It is possible to select more than one chromosome for hotspots. Play around.

![](choose_phenotype_11.png)

## Haplotype Scans

Switching the left (black) dashboard to the second entry (from `Phenotypes and Region` to `Haplotype Scan`) starts analyses. Note that most panels from here forward have `CSV` and `Plots` buttons on the side panel to save results.

![](lod_scan_1Mb.png)

The window looked too narrow, so switch back to `Phenotypes and Region` to change window width to 5Mb. The quickest way to do this is to grab the number in the `width` box with your mouse and type in a new value.

![](change_window.png)

And then switch back to see LOD scan.

![](lod_scan_5Mb.png)

Here is the LOD summary.

![](lod_scan_summary.png)

And here are the allele effects. Peak is indicated by vertical dashed line.

![](allele_effects.png)

We can get LOD and allele effects scan on same page.

![](lod_and_allele_effects.png)

#### SNP Scans

The next button in the side panel of the page is `SNP Association`.


![](SNP_scan.png)

We can zoom in to isolate the peak region using the slider. Sliders are present on sevaral of the panels. Be careful to make changes slowly, waiting for response.

![](SNP_scan_zoom.png)

Here is a summary table, which can be downloaded (see `CSV` button on lower left). This is a wide summary; you can scroll right to see more.

![](SNP_summary.png)

#### Genes and Exons

We can examine the genes in the peak region.

![](genes_in_region.png)

And underlay SNPs to see how they are related. Darker vertical lines indicate stronger SNP LOD scores.

![](genes_in_region_SNPs.png)

And zoom in to desired region using the slider again.

![](genes_in_region_SNPs_zoom.png)

Individual genes and their exons can be viewed as well.

![](gene_with_exons.png)

This `Exons` view also has its own summary table.

![](genes_summary.png)

#### Allele Pattern

SNP association seems rather complicated, but can be partially unraveled by grouping SNPs by their allele patterns. That is, SNPs take on two values (reference or alternate) for each allele. We can use that to color-code allele patterns.

![](allele_pattern.png)

And we can give that as a summary.

![](allele_pattern_summary.png)

Top SNPs can be identified by pattern ...

![](top_SNPs_pattern.png)

and by consequence.

![](top_SNPs_consequence.png)

#### Mediation

Mediation is a way to formally test if two phenotypes are causally related. Taking one phenotype as the `target`, we examine all other phenotypes that map to the same small region as possible `drivers`, and compare causal, reactive, independent and correlated models.

![](mediation.png)

In interactive mode (available for certain plots such as this), we can examine other information on tests.

![](mediation_interactive.png)

We can also get a summary for mediation

![](mediation_summary.png)

Behind the mediation is a subtle comparison of causal models. To get some handle on this for the data, check the `Scatter Plot` box and explore.

![](mediation_scatter.png)

## SNP/Gene Action

There are actually three possible values for a SNP: 0, 1, or 2 copies of the reference allele. We can use probabilities for these three values, derived from the genotype probabilities for allele pairs, to do SNP association with 2 degrees of freedom.

![](SNP_scan_dominance.png)

We can also examine allele patterns for additive and dominance effects.

![](allele_pattern_dominance.png)

Based on the best allele pattern, we can impute new SNP genotypes across the region that match this pattern, and construct LOD scans, now with 2 degrees of freedom per allele pattern.

![](lod_scan_dominance.png)

This might seem rather confusing, but it points out at least two interesting patterns that would otherwise be missed. The green line `ABDEH:CFG` has a peak at 97Mb, higher than where there were SNPs. The pink line for `ABDEFH:CG` imputes and extrapolates SNPs to the right of those found in the allele pattern plot above, revealing a peak at 101Mb. These may or may not be real, but at least provide a cautionary tale.
 
There are several possible gene actions that could be explored.

![](gene_action_menu.png)

## Multiple Phenotypes

It is very easy to look at multiple phenotypes together. At the `Phenotypes and Region`, just select more than one. [Warning that if you select more than five, you are likely to see a serious slowdown, as all calculations are done in real time!] In some cases, multiple phenotypes will appear on the same figure, possibly as multiple facets; in other cases, there will be a choice box in the side panel to switch among phenotypes. Here are just a few examples.

![](multiple_choose.png)

![](multiple_lod.png)

![](multiple_lod_allele.png)

![](multiple_SNP.png)
