---
title: "QTL2 Shiny Interface"
author: "Brian S. Yandell"
date: "1/12/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This document concerns a suite of packages complementary to [R/qtl2](http://www.rqtl.org/qtl2) that extend capabilities in terms of plotting, fast data access, deeper probe of SNPs and structural variants, and a user-friendly interface for deep investigation of small intervals. 

The QTL2 Shiny Interface can be invoked from the Shiny server
<http://www.statlab.wisc.edu/shiny/qtl2shiny>. We demonstrate with a publicly available dataset from Recla (see <https://github.com/rqtl/qtl2data>).
However, this Shiny server is set up
with project-specific login and password protection for data that has not been published, using [server side Apache htaccess](https://httpd.apache.org/docs/current/howto/ssi.html).

## Strategy to Study Small Interval

Here is a strategy to look at multiple traits in a small interval. These are rough and ready notes that will be smoothed out in time.

- use analyses table (or other approach) to identify region of interest
    + say 1-5Mb window
- identify set of traits in region
- estimate allele effects at peak for each trait
    + useful plots?
- find best SNP(s) and pattern(s)
    + identify SDP and pattern for each
    + useful plots?
- do same for full effects
    + 36 diplotypes
    + 3-level SNPs
    
Some challenges

- don't have useful thresholds yet
- would like to think about inference across related patterns
- multiple peaks in small region
- pleiotropy vs close linkage

How to incorporate this into shiny tool?

- quick computation: limit size of region
- do we have to run genome scan?
    + or just consider flanking markers for each peak?

## QTL2 Shiny Interface

Information will appear here about how the Shiny Interface works. Or perhaps that is best as another package. 

Most of the rest of this is more of a Developer Guide, to explain the data structures and how the [R/qtl2shiny](https://github.com/byandell/qtl2shiny) package is organized. Those wishing to _use_ the shiny interface with their own data will want to read about the data structures, and contact me. Those wishing to _extend_ capabilities of the package, or understand its inner working, with be interested in the package organization.

## R/qtl2 companion packages

The popular [R/qtl](http://www.rqtl.org) package for gene mapping has been redeveloped for high volume, multi-parent systems genetics data as [R/qtl2](http://www.rqtl.org/qtl2). This new package is still in a state of flux, so please work with us to improve.

This document describes companion packages that are used to build the Shiny server:

- [R/qtl2ggplot2](https://github.com/byandell/qtl2ggplot2)
- [R/qtl2feather](https://github.com/byandell/qtl2feather)
- [R/qtl2pattern](https://github.com/byandell/qtl2pattern)
- [R/qtl2shiny](https://github.com/byandell/qtl2shiny)

The three other packages can be employed on their own separate from the server. In particular, [R/qtl2ggplot2](https://github.com/byandell/qtl2ggplot2) provides full capability for `ggplot2` plot objects. The [R/qtl2feather](https://github.com/byandell/qtl2feather) package transforms large, slow `calc_genoprob` objects central to [R/qtl2](http://www.rqtl.org/qtl2) into a set of `feather` databases for quick access. See each package for its own documentation. 

There are two additional packages that are more specialized and still need documentation:

- [R/DOread](https://github.com/byandell/DOread)
- [R/CausalMST](https://github.com/byandell/CausalMST)

[R/DOread](https://github.com/byandell/DOread) has code to set up Diversity Outbred (DO) mouse data for the Shiny server, and routines to read those data.
[R/CausalMST](https://github.com/byandell/CausalMST) concerns Causal Model Selection Tests, which are used in conjuction with mediation to formally infer causal relationship between a phenotype and other phenotypes that co-map.
    
## Organization of qtl2shiny data files

The interface assumes the Shiny app file [app.R](https://github.com/byandell/qtl2shiny/blob/master/inst/qtl2shiny/app.R) is in a directory with other support documents and the data. The projects are identified in a CSV file `projects.csv`, which has the following content:

```
project,taxa,directory
Recla,CCmouse,qtl2shinyData
```

The column names are fixed. The `project` is the project name (`Recla`), `taxa` has the taxa name (`CCmouse`), and `directory` is the relative or absolute address of the directory where data are kept (`qtl2shinyData`, which is relative here). In this case, the basic structure of the app is:

```
app.R
about.md
about-extended.md
project.csv
qtl2shinyData/
  CCmouse/
    Recla/
```

The `about.md` and `about-extended.md` files are place-holders for now for online documentation. All data are now stored under the directory (`qtl2shinyData`) using relative addresses. However, it is possible for data for a `taxa` and `project` to be stored in another place, in which case an absolute address would be supplied in the `projects.csv` file.

Detailed information about the `taxa` and `project` can be found with the [CCmouse.Rmd](https://github.com/byandell/qtl2shiny/blob/master/inst/doc/CCmouse.Rmd) and [Recla.Rmd](https://github.com/byandell/qtl2shiny/blob/master/inst/doc/Recla.Rmd) examples, respectively.

### Taxa information

The `taxa` sub-directory (`CCmouse`) contains the following files. See <https://github.com/rqtl/qtl2db>
for explanation of the SQL databases and query routines.

file | contents
-----|---------
`allele_info.rds` | information on alleles for taxa
`cc-variants.sqlite` | CC structural variants in SQL database
`mouse_genes_mgi.sqlite` | mouse genes curated by MGI in SQL database
`query_genes.rds` | query routine for genes
`query_variants.rds` | query routines for structural variants

### Project Information

The `project` sub-sub-directory (`Recla`) has the data specific to the project. An example of how this is created can be found in [Recla.Rmd](https://github.com/byandell/qtl2shiny/blob/master/inst/doc/Recla.Rmd).
Most of the objects are stored as [RDS](https://www.rdocumentation.org/packages/base/versions/3.4.3/topics/readRDS).
The genotype probabilities are converted from `calc_genoprob` to `feather_genoprob` format using [R/qtl2feather](https://github.com/byandell/qtl2feather).
Query routines are somewhat analogous to those used for `taxa` above.

Note that [R/qtl2](http://www.rqtl.org/qtl2) wants data files to be consistent in terms of individual identifiers. The `covar.rds` and `pheno_data.rds` files have identifiers as row names, which are also used in the genotype probabilities.

file | contents
-----|------------------------------------------------
`genoprob` | directory with genotype probabilities created with [R/qtl2feather](https://github.com/byandell/qtl2feather)
`kinship.rds` | kinship object using LOCO (see [R/qtl2](http://kbroman.org/qtl2/assets/vignettes/user_guide.html#calculating_a_kinship_matrix))
`pmap.rds` | physical map
`analyses.rds` | data frame of information about analyses (one row per phenotype)
`covar.rds` | data frame of covariates (row = individual, column = covariate)
`pheno_data.rds` | matrix of phenotype data (row = individual, column = phenotype)
`pheno_type.rds` | vector of phenotype types (this will probably disappear)
`peaks.rds` | data frame of peak information (one row per peak)
`hotspot.rds` | hotspot object with counts of peaks by position and phenotype set
`query_mrna.rds` | query of mRNA data
`query_probs.rds` | query of genotype probabilities

Likely changes:

- pheno_type.rds will disappear
- pheno_data.rds may become a feather database
- analyses.rds will be redesigned

Some projects have an `RNAseq` folder. This, and accompanying routines for mRNA now in [R/DOread](https://github.com/byandell/DOread) and other packages, will likely be redesigned. Currently, code is set up for only one type of mRNA data.

## Data files


## Query files

The query files are designed to encapsulate specific information about where data are stored and how to retrieve them, so that different projects might use different strategies. For instance, some projects might use CSV or RDS files, while others might use SQL or feather databases. There are four query objects, for genes, variants, mrna and probs.

Note that if you use a similar format as [CCmouse.Rmd](https://github.com/byandell/qtl2shiny/blob/master/inst/doc/CCmouse.Rmd) and [Recla.Rmd](https://github.com/byandell/qtl2shiny/blob/master/inst/doc/Recla.Rmd) examples
to create and store the queries, you may have to quit RStudio and restart it to have it properly pick up the `project` or `taxa` directory.

The following two `create` functions are in [R/qtl2](http://www.rqtl.org/qtl2). They create functions (`query_genes` and `query_variants`) that contain, in their local environment, the location of the respective SQL database. Saving these as RDS preserves this environment information for later reuse.

```{r eval=FALSE}
query_genes <- 
  qtl2::create_gene_query_func("CCmouse/mouse_genes.sqlite")
saveRDS(query_genes, "query_genes.rds")
```

```{r eval=FALSE}
query_variants <- 
  qtl2::create_variant_query_func("CCmouse/cc_variants.sqlite"))
saveRDS(query_variants, "query_variants.rds")
```

This allows calls of the following form to be embedded in code, without reference to where the database is, or even what type of database it might be. Here we look on chromosome `"1"` between `39` and `40` Mbp.

```{r eval=FALSE}
genes <- query_genes("1", 39, 40)
variants <- query_variants("1", 39, 40)
```

The above two routines would generally be used across all projects involving the CC mouse. The following two routines are project specific, refering to genotype probabilities and mRNA data. They do not directly identify the data file, as that is currently hard-coded into the [R/DOread](https://github.com/byandell/DOread) package. These will change substantially in the near future.

```{r}
query_probs <- DOread::create_probs_query_func_do("CCmouse/Recla")
saveRDS(query_probs, "query_probs.rds")
```

```{r}
query_mrna <- DOread::create_probs_query_func_do("CCmouse/Recla")
saveRDS(query_probs, "query_mrna.rds")
```

Again, these enable calls to access data in a region. I include additional arguments that are used for these routines. The call, respectively, `DOread::read_probs` and `DOread::read_mrna`.

```{r eval=FALSE}
probs <- query_probs("1", 39, 40, allele = TRUE, method = "feather")
mrna <- query_mrna("1", 39, 40, local = TRUE, qtl = FALSE)
```


## Details of Data and Shiny Setup

The `qtl2shiny` interface can be invoked within R using [runApp](https://shiny.rstudio.com/reference/shiny/latest/runApp.html) on the file
[app.R](https://github.com/byandell/qtl2shiny/blob/master/inst/qtl2shiny/app.R). There is a routine [qtl2shinyApp](https://github.com/byandell/qtl2shiny/blob/master/R/qtl2shinyApp.R) to do this as well. However, one first should set up the data and query files.