---
title: "QTL2 Shiny Interface"
author: "Brian S. Yandell"
date: "1/12/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The popular [R/qtl](http://www.rqtl.org) package for gene mapping has been redeveloped for high volume, multi-parent systems genetics data as [R/qtl2](http://www.rqtl.org/qtl2). This new package is still in a state of flux, so please work with us to improve.

This document concerns a suite of packages complementary to [R/qtl2](http://www.rqtl.org/qtl2) that extend capabilities in terms of plotting, fast data access, deeper probe of SNPs and structural variants, and a user-friendly interface for deep investigation. The packages are:

- [R/qtl2ggplot2](https://github.com/byandell/qtl2ggplot2)
- [R/qtl2feather](https://github.com/byandell/qtl2feather)
- [R/qtl2pattern](https://github.com/byandell/qtl2pattern)
- [R/qtl2shiny](https://github.com/byandell/qtl2shiny)

See the other packages for their own documentation. 
There are two additional packages that are more specialized and still need documentation:

- [R/DOread](https://github.com/byandell/DOread)
- [R/CausalMST](https://github.com/byandell/CausalMST)

## QTL2 Shiny Interface

The QTL2 Shiny Interface can be invoked from the Shiny server
<http://www.statlab.wisc.edu/shiny/qtl2shiny>. This is being set up
with project-specific login and password protection for data that has not been published, using [server side Apache htaccess](https://httpd.apache.org/docs/current/howto/ssi.html).

## Organization of taxa and project files

The interface assumes the Shiny app file [app.R](https://github.com/byandell/qtl2shiny/blob/master/inst/qtl2shiny/app.R) is in a directory with other support documents and the data. The projects are identified in a CSV file `projects.csv`, which has the following content:

```
project,taxa,directory
Recla,CCmouse,qtl2shinyData
```

The column names are fixed. The `project` is the project name (`Recla`), `taxa` has the taxa name (`mouse`), and `directory` is the relative or absolute address of the directory where data are kept (`qtl2shinyData`, which is relative here). In this case, the basic structure of the app is:

```
app.R
about.md
about-extended.md
project.csv
qtl2shinyData/
  mouse/
    Recla/
```

The `taxa` sub-directory (`mouse`) contains the following files. See <https://github.com/rqtl/qtl2db>
for explanation of the SQL databases and query routines.

file | contents
-----|---------
allele_info.rds | information on alleles for taxa
cc-variants.sqlite | CC structural variants in SQL database
mouse_genes_mgi.sqlite | mouse genes curated by MGI in SQL database
query_genes.rds | query routine for genes
query_variants.rds | query routines for structural variants

The `project` sub-sub-directory (`Recla`) has the data specific to the project. An example of how this is created can be found in [Recla.Rmd](https://github.com/byandell/qtl2shiny/blob/master/inst/doc/Recla.Rmd).
Most of the objects are stored as [RDS](https://www.rdocumentation.org/packages/base/versions/3.4.3/topics/readRDS).
The genotype probabilities are converted from `calc_genoprob` to `feather_genoprob` format using [R/qtl2feather](https://github.com/byandell/qtl2feather).
Query routines are somewhat analogous to those used for `taxa` above.

Note that [R/qtl2](http://www.rqtl.org/qtl2) wants data files to be consistent in terms of individual identifiers. The `covar.rds` and `pheno_data.rds` files have identifiers as row names, which are also used in the genotype probabilities.

file | contents
-----|------------------------------------------------
genoprob | directory with genotype probabilities created with [R/qtl2feather](https://github.com/byandell/qtl2feather)
kinship.rds | kinship object using LOCO (see [R/qtl2](http://kbroman.org/qtl2/assets/vignettes/user_guide.html#calculating_a_kinship_matrix))
pmap.rds | physical map
analyses.rds | data frame of information about analyses (one row per phenotype)
covar.rds | data frame of covariates (row = individual, column = covariate)
pheno_data.rds | matrix of phenotype data (row = individual, column = phenotype)
pheno_type.rds | vector of phenotype types (this will probably disappear)
peaks.rds | data frame of peak information (one row per peak)
hotspot.rds | hotspot object with counts of peaks by position and phenotype set
query_mrna.rds | query of mRNA data
query_probs.rds | query of genotype probabilities

## Data files


## Query files

The query files are designed to encapsulate specific information about where data are stored and how to retrieve them, so that different projects might use different strategies. For instance, some projects might use CSV or RDS files, while others might use SQL or feather databases. There are four query objects, for genes, variants, mrna and probs.


## Details of Data and Shiny Setup

The `qtl2shiny` interface can be invoked within R using [runApp](https://shiny.rstudio.com/reference/shiny/latest/runApp.html) on the file
[app.R](https://github.com/byandell/qtl2shiny/blob/master/inst/qtl2shiny/app.R). There is a routine [qtl2shinyApp](https://github.com/byandell/qtl2shiny/blob/master/R/qtl2shinyApp.R) to do this as well. However, one first should set up the data and query files.