---
title: "recla"
author: "Brian S. Yandell"
date: "12/8/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
project_info <- data.frame(project = "Recla",
                           taxa = "CCmouse",
                           directory = "qtl2shinyData",
                           stringsAsFactors = FALSE)
```

```{r}
project_dir <- project_info$project
if(!dir.exists(project_dir)) {
  dir.create(project_dir)
}
```

```{r message=FALSE}
library(dplyr)
library(qtl2)
library(qtl2feather)
library(qtl2pattern)
```

```{r}
file <- paste0("https://raw.githubusercontent.com/rqtl/",
               "qtl2data/master/DO_Recla/recla.zip")
recla <- read_cross2(file)
```

```{r}
saveRDS(recla$pmap, file.path(project_dir, "pmap.rds"))
```

```{r}
covar <- recla$covar %>%
  rename(sex = Sex)
saveRDS(covar, file.path(project_dir, "covar.rds"))
```

```{r}
saveRDS(recla$pheno, file.path(project_dir, "pheno_data.rds"))
```

## Genotype probabilities

```{r}
if(!dir.exists(feather_dir <- file.path(project_dir, "genoprob"))) {
  dir.create(feather_dir)
}
faprobs_file <- file.path(feather_dir, "faprobs.rds")
if(!file.exists(probs_file <- file.path(feather_dir, "fprobs.rds"))) {
  map <- insert_pseudomarkers(recla$gmap, step=1)
  
  cat("allele pair genotype probabilities\n", file = stderr())
  pr <- calc_genoprob(recla, map, err=0.002)
  cat("allele pair prob conversion via qtl2feather\n", file = stderr())
  fprobs <- feather_genoprob(pr, "probs", feather_dir, verbose = FALSE)
  saveRDS(fprobs, file = file.path(feather_dir, "fprobs.rds"))

  cat("allele genotype probabilities\n", file = stderr())
  apr <- genoprob_to_alleleprob(pr)
  cat("allele prob conversion via qtl2feather\n", file = stderr())
  faprobs <- feather_genoprob(apr, "aprobs", feather_dir, verbose = FALSE)
  saveRDS(faprobs, file = faprobs_file)
} else {
  faprobs <- readRDS(faprobs_file)
}
```

```{r}
kinship_loco <- calc_kinship(pr, "loco")
```

```{r}
saveRDS(kinship_loco, file.path(project_dir, "kinship.rds"))
```

## Genome Scans

```{r}
form <- formula(paste("~", paste(names(covar), collapse = "+")))
addcovar <- model.matrix(form, covar)[, -1, drop = FALSE]
out <- scan1(faprobs, recla$pheno, addcovar=addcovar)
```

```{r}
peaks <- find_peaks(out, recla$pmap, threshold = 3)
```

## Set up analyses and peaks tables for qtl2shiny

```{r}
peaks <- peaks %>%
  rename(
    pheno = lodcolumn) %>%
  mutate(
    longname = pheno,
    output = pheno,
    pheno_group = "recla",
    pheno_type = "recla") %>%
  select(-lodindex)
```

```{r}
saveRDS(peaks, file.path(project_dir, "peaks.rds"))
```

```{r}
analyses_tbl <- distinct(peaks, pheno, .keep_all = TRUE) %>%
  select(-lod, -pos, -chr) %>%
  mutate(model = "normal",
         transf = "identity",
         offset = 0,
         winsorize = FALSE)
for(i in names(covar))
  analyses_tbl[[i]] <- TRUE
```

```{r}
saveRDS(analyses_tbl, file.path(project_dir, "analyses.rds"))
```

```{r}
pheno_type   <- DOread::setup_type(analyses_tbl)
```

```{r}
saveRDS(pheno_type, file.path(project_dir, "pheno_type.rds"))
```

## Hotspots

```{r}
hots <- qtl2pattern::hotspot(recla$pmap, peaks)
```

```{r}
saveRDS(hots, file.path(project_dir, "hotspot.rds"))
```

## Query functions

The query for genes and variants are usually common across all projects for a taxa.
Here we use query routines from R/qtl2 for the CC mouse.

```{r}
query_genes <- 
  qtl2::create_gene_query_func(
    file.path("qtl2shinyData", "CCmouse", "mouse_genes.sqlite"))
saveRDS(query_genes, file.path(project_dir, "..", "query_genes.rds"))
```

```{r}
query_variants <- 
  qtl2::create_variant_query_func(
    file.path("qtl2shinyData", "CCmouse", "cc_variants.sqlite"))
saveRDS(query_variants, file.path(project_dir, "..", "query_variants.rds"))
```

Use `create_probs_query_func_do` from package `DOread` to query genotype probabilities.

```{r}
project_path <- file.path(project_info$directory,
              project_info$taxa,
              project_info$project)
query_probs <- 
  DOread::create_probs_query_func_do(
    project_path)
saveRDS(query_probs, file.path(project_dir, "query_probs.rds"))
```

These data do not include mRNA data, so set up null routine. Make sure it has
required arguments.

```{r}
tmpfn <- function(dbfile) {
  function(chr, start, stop,
           local = TRUE,
           qtl = FALSE) NULL
}
query_mrna <- tmpfn()
saveRDS(query_mrna, file.path(project_dir, "query_mrna.rds"))
```


