---
title: "recla"
author: "Brian S. Yandell"
date: "12/8/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This sets up the `Recla` project using data from <https://github.com/rqtl/qtl2data> for use with [R/qtl2shiny](https://github.com/byandell/qtl2shiny).
See also [CCmouse.Rmd](https://github.com/byandell/qtl2shiny/blob/master/inst/doc/CCmouse.Rmd) for needed files for `CCmouse` taxa.

The `project_info` data frame contains control information about each project (see [R/qtl2shiny](https://github.com/byandell/qtl2shiny)).

```{r}
project_info <- data.frame(project = "Recla",
                           taxa = "CCmouse",
                           directory = "qtl2shinyData",
                           stringsAsFactors = FALSE)
```

Assume local directory already created with `taxa` and `project` under.

```{r}
localdir <- "../../.."
```

```{r}
(project_dir <- file.path(localdir, project_info$taxa, project_info$project))
```

```{r}
if(!dir.exists(project_dir)) {
  warning("project director does not exist")
  knitr::knit_exit()
}
```

## Query functions

Use `create_probs_query_func_do` from package `qtl2pattern` to query genotype probabilities.

```{r}
query_probs <- qtl2pattern::create_probs_query_func("qtl2shinyData/CCmouse/Recla")
saveRDS(query_probs, file.path(project_dir, "query_probs.rds"))
```

These data do not include mRNA data, so set up null routine. Make sure it has
required arguments.

```{r}
query_mrna <- qtl2pattern::create_mrna_query_func(NULL)
saveRDS(query_mrna, file.path(project_dir, "query_mrna.rds"))
```

```{r}
if(!dir.exists(project_dir)) {
  if(!dir.exists(project_info$taxa))
    dir.create(project_info$taxa)
  dir.create(project_dir)
}
```

```{r message=FALSE}
library(dplyr)
library(qtl2)
library(qtl2feather)
library(qtl2pattern)
```

```{r}
recla <- 
  read_cross2(
    paste0(
      "https://raw.githubusercontent.com/rqtl/",
      "qtl2data/master/DO_Recla/recla.zip"))
```

```{r}
covar <- recla$covar %>%
  rename(sex = Sex)
```

```{r}
saveRDS(recla$pmap, file.path(project_dir, "pmap.rds"))
saveRDS(covar, file.path(project_dir, "covar.rds"))
saveRDS(recla$pheno, file.path(project_dir, "pheno_data.rds"))
```

## Genotype probabilities

```{r}
if(!dir.exists(feather_dir <- file.path(project_dir, "genoprob"))) {
  dir.create(feather_dir)
}
faprobs_file <- file.path(feather_dir, "faprobs.rds")
if(!file.exists(fprobs_file <- file.path(feather_dir, "fprobs.rds"))) {
  map <- insert_pseudomarkers(recla$gmap, step=1)
  
  cat("allele pair genotype probabilities\n", file = stderr())
  pr <- calc_genoprob(recla, map, err=0.002)
  cat("allele pair prob conversion via qtl2feather\n", file = stderr())
  fprobs <- feather_genoprob(pr, "probs", feather_dir, verbose = FALSE)
  saveRDS(fprobs, file = fprobs_file)

  cat("allele genotype probabilities\n", file = stderr())
  apr <- genoprob_to_alleleprob(pr)
  cat("allele prob conversion via qtl2feather\n", file = stderr())
  faprobs <- feather_genoprob(apr, "aprobs", feather_dir, verbose = FALSE)
  saveRDS(faprobs, file = faprobs_file)
} else {
  fprobs <- readRDS(fprobs_file)
  faprobs <- readRDS(faprobs_file)
}
```

```{r}
kinship_loco <- calc_kinship(faprobs, "loco")
```

```{r}
saveRDS(kinship_loco, file.path(project_dir, "kinship.rds"))
```

## Genome Scans

```{r}
form <- formula(paste("~", paste(names(covar), collapse = "+")))
addcovar <- model.matrix(form, covar)[, -1, drop = FALSE]
out <- scan1(faprobs, recla$pheno, addcovar=addcovar)
```

```{r}
peaks <- find_peaks(out, recla$pmap, threshold = 3)
```

## Set up analyses and peaks tables for qtl2shiny

```{r}
peaks <- peaks %>%
  rename(
    pheno = lodcolumn) %>%
  mutate(
    longname = pheno,
    output = pheno,
    pheno_group = "Recla",
    pheno_type = "recla") %>%
  select(-lodindex)
```

```{r}
analyses_tbl <- distinct(peaks, pheno, .keep_all = TRUE) %>%
  select(-lod, -pos, -chr) %>%
  mutate(model = "normal",
         transf = "identity",
         offset = 0,
         winsorize = FALSE)
for(i in names(covar))
  analyses_tbl[[i]] <- TRUE
```

```{r}
saveRDS(peaks, file.path(project_dir, "peaks.rds"))
saveRDS(analyses_tbl, file.path(project_dir, "analyses.rds"))
```

## Hotspots

```{r}
hots <- qtl2pattern::hotspot(recla$pmap, peaks)
```

```{r}
saveRDS(hots, file.path(project_dir, "hotspot.rds"))
```
