---
title: "Recla qtl2shiny Demo"
author: "Brian S. Yandell"
date: "`r format(Sys.Date(), '%d %b %Y')`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This is a demo of the Recla et al. dataset using functionality of [R/qtl2shiny](https://github.com/byandell/qtl2shiny) shown in <http://www.stat.wisc.edu/~yandell/software/qtl2shiny>. We assume that the 
[CCmouse.Rmd](https://github.com/byandell/qtl2shiny/blob/master/inst/doc/CCmouse.Rmd) and 
[Recla.Rmd](https://github.com/byandell/qtl2shiny/blob/master/inst/doc/Recla.Rmd)
scripts have already been run to put data in the `CCmouse` directory.

Here we just attach the `dplyr` and `ggplot2` packages, but explicitly use
elements of the `qtl2`, `qtl2ggplot` and `qtl2pattern` packages. 
The `qtl2fst` package is used implicitly for access to genotype probabilities.

```{r message=FALSE}
library(dplyr)
library(ggplot2)
```

```{r}
project_info <- data.frame(project = "Recla",
                           taxa = "CCmouse",
                           directory = "../../..",
                           stringsAsFactors = FALSE)
```

```{r}
(taxa_dir <- file.path(project_info$directory, 
                          project_info$taxa))
```

```{r}
(project_dir <- file.path(taxa_dir, project_info$project))
```

Assume local directory already created with `taxa` and `project` under,
and with all data already installed.

```{r}
if(!dir.exists(project_dir)) {
  warning("project director does not exist")
  knitr::knit_exit()
}
```

List files for the taxa and for the project.

```{r}
list.files(project_info$taxa)
```

```{r}
list.files(project_dir)
```

```{r}
list.files(file.path(project_dir, "genoprob")) 
```

## Choose Region and Phenotype

```{r}
# See shinyHotspot.R
hotspots <- readRDS(file.path(project_dir, "hotspot.rds"))
```

```{r}
# See shinyHotspot.R
qtl2ggplot::ggplot_scan1(hotspots$scan, hotspots$map)
```

```{r}
# See shinyHotspot.R
(hot_sum <- qtl2pattern::summary_scan1(hotspots$scan, hotspots$map) %>%
  filter(pheno == "all") %>%
  rename(count = lod) %>%
  select(-marker) %>%
  arrange(desc(count)))
```

```{r}
chr_id <- as.character(hot_sum$chr[1])
pos_Mbp <- hot_sum$pos[1]
window_Mbp <- 5
start_val <- pos_Mbp - window_Mbp
end_val <- pos_Mbp + window_Mbp
```

```{r}
# See shinyHotspot.R
autoplot(hotspots$scan, hotspots$map, chr = chr_id)
```

```{r}
# See shinyMain.R
peaks <- readRDS(file.path(project_dir, "peaks.rds")) %>%
  filter(chr == chr_id,
         pos >= start_val,
         pos <= end_val)
```

```{r}
# See shinyPhenos.R
(peaks <- peaks %>%
  select(pheno, chr, pos, lod) %>%
  arrange(desc(lod)))
```

```{r}
pheno_name <- peaks$pheno[1]
```

#### Phenotype Data

```{r}
# See shinyMain.R
(analyses <- readRDS(file.path(project_dir, "analyses.rds")) %>%
  filter(pheno == pheno_name))
```

```{r}
# See shinyMain.R
(pheno_data <- readRDS(file.path(project_dir, "pheno_data.rds")) %>%
   qtl2pattern::pheno_trans(analyses$pheno, 
               analyses$transf,
               analyses$offset,
               analyses$winsorize)) %>% 
  head
```

```{r}
# See shinyMain.R
(covar <- readRDS(file.path(project_dir, "covar.rds")) %>%
  qtl2pattern::get_covar(analyses)) %>%
  head
```

## Genome Scan

```{r}
# See shinyMain.R
kinship <- readRDS(file.path(project_dir, "kinship.rds"))[chr_id]
```

```{r}
# See shinyProbs.R and Recla.Rmd
query_probs <- qtl2pattern::create_probs_query_func(project_dir)
probs_obj <- query_probs(chr_id, start_val, end_val)
```

```{r}
names(probs_obj)
dim(probs_obj$probs)
```

```{r}
str(probs_obj$map)
```

```{r}
# See shinyScanCoef.R
scan_obj <- qtl2pattern::scan1_covar(pheno_data, covar, probs_obj$probs, kinship,
                        analyses)
```

```{r}
autoplot(scan_obj, probs_obj$map)
```

```{r}
summary(scan_obj, probs_obj$map)
```

```{r}
qtl2::find_peaks(scan_obj, probs_obj$map)
```

```{r}
addcovar <- qtl2pattern::covar_df_mx(covar)
```

```{r}
eff_obj <- qtl2::scan1coef(probs_obj$probs, pheno_data, kinship, addcovar)
```

```{r}
autoplot(eff_obj, probs_obj$map) +
  geom_vline(xintercept = peaks$pos[1], linetype = "dashed")
```


```{r}
autoplot(eff_obj, probs_obj$map, scan1_output = scan_obj,
         legend.position = "none")
```

## SNP Scans

```{r}
# See shinySNPSetup.R and shinyProbs.R
query_variants <- qtl2::create_variant_query_func(
  file.path(taxa_dir, "cc_variants.sqlite"))
snpinfo <- query_variants(chr_id, start_val, end_val)
```

```{r}
snpprobs_obj <- qtl2pattern::get_snpprobs(chr_id, pos_Mbp, window_Mbp,
                   pheno_name, 
                   probs_obj$probs,
                   probs_obj$map,
                   snpinfo)
```

```{r}
snp_scan_obj <- qtl2pattern::scan1_covar(pheno_data, covar, snpprobs_obj$snpprobs, kinship, analyses)
```

```{r}
autoplot(snp_scan_obj, snpprobs_obj$snpinfo, show_all_snps = FALSE,
         drop_hilit = 1.5,)
```

```{r}
# See shinySNPSum and shinySNPSetup
top_snps_tbl <- qtl2pattern::top_snps_all(snp_scan_obj,
                        snpprobs_obj$snpinfo,
                        1.5)
(patterns <- summary(top_snps_tbl))
```

#### Genes and Exons

```{r}
# See shinyGeneRegion.R
query_genes <- qtl2::create_gene_query_func(
  file.path(taxa_dir, "mouse_genes.sqlite"))
rng <- range(top_snps_tbl$pos) + c(-.005, .005)
gene_tbl <- query_genes(chr_id, rng[1], rng[2])
gene_region_tbl <- qtl2pattern::get_genes(chr_id, rng[1], rng[2], gene_tbl)
```

```{r}
autoplot(gene_region_tbl)
```

```{r}
autoplot(gene_region_tbl, top_snps_tbl = top_snps_tbl)
```

```{r}
# See shinyGeneExon.R
gene_exon_tbl <- qtl2pattern::get_gene_exon_snp(top_snps_tbl, gene_tbl)
```

```{r}
(exon_sum <- summary(gene_exon_tbl, top_snps_tbl = top_snps_tbl))
```

```{r}
autoplot(gene_exon_tbl, top_snps_tbl, genes = exon_sum$gene[1])
```

#### SNP Allele Patterns

```{r}
# See shinySNPPattern.R
autoplot(snp_scan_obj, snpprobs_obj$snpinfo, patterns = "hilit", drop_hilit = 1.5)
```

```{r}
# See shinySNPFeature.R
top_feature <- qtl2pattern::merge_feature(top_snps_tbl, snpprobs_obj$snpinfo,
                             snp_scan_obj, 1.5, 0, gene_exon_tbl)
```

```{r}
autoplot(top_feature, pheno_name, "pattern")
```

```{r}
autoplot(top_feature, pheno_name, "consequence")
```

#### Mediation

This section (and package `CausalMST`) is working but will likely be rearranged.

First, identify all phenotypes that map to the region. This is done using `peaks` in the region to identify phenotype names and `all_analyses` to construct the appropriate phenotype transformation.

```{r}
# See shinyMediate.R
all_analyses <- readRDS(file.path(project_dir, "analyses.rds")) %>%
  filter(pheno != pheno_name,
         pheno %in% peaks$pheno)
all_pheno_data <- readRDS(file.path(project_dir, "pheno_data.rds")) %>%
   qtl2pattern::pheno_trans(all_analyses$pheno, 
               all_analyses$transf,
               all_analyses$offset,
               all_analyses$winsorize)
```

The following works for mediators as other phenotypes. But see `qtl2pattern::expr_region()` and `qtl2pattern::create_mrna_query_func()` to create `query_mrna()` to access expression data.

The basic idea here is to construct a list containing at least three items:

- `pheno`: matrix of phenotypes (columns) by individuals (rows)
- `annot`: data frame of annotation
- `covar`: data frame of covariates
- `peaks`: data frame of peak information (optional)

```{r}
med_ls <- qtl2pattern::pheno_region(
  chr_id, start_val, end_val, covar, probs_obj$map,
  peaks, all_analyses, all_pheno_data)
```

```{r}
peak_mar <- qtl2::find_marker(probs_obj$map, chr_id, pos_Mbp)
# waiting for fix to qtl2
#geno_max <- qtl2::pull_genoprobpos(probs_obj$probs, peak_mar)
geno_max <- subset(probs_obj$probs, mar = peak_mar)[[1]][,,1]
```

Mediate the target (`pheno_data`) using information in `med_ls` about potential mediators. The driver for the target is `geno_max`, while the driver for the mediators come from `driver_med` using names from `med_ls$annot$driver`.

```{r}
mediate_obj <- CausalMST::mediation_test(
  target = pheno_data,
  mediator = med_ls[[1]], 
  annotation = med_ls[[2]],
  driver = geno_max, 
  covar_tar = covar,
  covar_med = med_ls$covar,
  kinship = kinship[[1]],
  driver_med = probs_obj$probs[[chr_id]],
  test = "wilc",
  pos = pos_Mbp)
```

```{r}
autoplot(mediate_obj)
```

Now look at one of the mediators (the strongest one) in a scatter plot against the target. There are various flavors of plot for this.

```{r}
scat_ls <- qtl2pattern::pull_mediator(mediate_obj, med_ls, patterns)
```

```{r}
scat_dat <- CausalMST::med_scatter(geno_max, pheno_data, 
                                   scat_ls$mediator,
                                   kinship[[1]], 
                                   covar, 
                                   med_ls$covar,
                                   qtl2::fit1,
                                   sdp = scat_ls$sdp, 
                                   allele = TRUE)
```

```{r}
autoplot(scat_dat)
```

## Gene Action

```{r}
# See shinyDiplo.R and shinyProbs.R
pairprobs_obj <- query_probs(chr_id, start_val, end_val, allele = FALSE)
```

```{r}
snppairprobs_obj <- qtl2pattern::get_snpprobs(chr_id, pos_Mbp, window_Mbp,
                   pheno_name, 
                   pairprobs_obj$probs,
                   pairprobs_obj$map,
                   snpinfo)
```

```{r}
snppair_scan_obj <- qtl2pattern::scan1_covar(pheno_data, covar, 
                                snppairprobs_obj$snpprobs, kinship, analyses)
```

```{r}
autoplot(snppair_scan_obj, snppairprobs_obj$snpinfo, show_all_snps = FALSE,
         drop_hilit = 1.5,)
```

```{r}
# See shinySNPPattern.R
autoplot(snppair_scan_obj, snppairprobs_obj$snpinfo, patterns = "hilit", drop_hilit = 1.5)
```

```{r}
top_snppairs_tbl <- 
  qtl2pattern::top_snps_all(snppair_scan_obj,
               snppairprobs_obj$snpinfo,
               1.5)
```

```{r}
(patterns <- summary(top_snppairs_tbl))
```

```{r}
# See shinyPattern.R
scan_pat <- qtl2pattern::scan_pattern(pairprobs_obj$probs,
                            pheno_data,
                            kinship, addcovar,
                            pairprobs_obj$map,
                            patterns)
```

```{r}
pattern_choices <- qtl2pattern::sdp_to_pattern(patterns$sdp, LETTERS[1:8])
```

```{r}
autoplot(scan_pat, pairprobs_obj$map, "lod", pattern_choices)
```

## Multiple Phenotypes

```{r}
pheno_names <- peaks$pheno[1:2]
```

Assume for now covariates are the same. The routine `qtl2pattern::scan1_covar` helps sort out covariates that change among phenotypes.

```{r}
# See shinyMain.R
(analyses <- readRDS(file.path(project_dir, "analyses.rds")) %>%
  filter(pheno %in% pheno_names))
```

```{r}
# See shinyMain.R
(pheno_data <- readRDS(file.path(project_dir, "pheno_data.rds")) %>%
   qtl2pattern::pheno_trans(analyses$pheno, 
               analyses$transf,
               analyses$offset,
               analyses$winsorize)) %>% 
  head
```
```{r}
# See shinyScanCoef.R
scan_obj <- qtl2pattern::scan1_covar(pheno_data, covar, probs_obj$probs, kinship,
                        analyses)
```

```{r}
autoplot(scan_obj, probs_obj$map, 1:2)
```

```{r}
snp_scan_obj <- qtl2pattern::scan1_covar(pheno_data, covar, snpprobs_obj$snpprobs, kinship, analyses)
```

```{r}
autoplot(snp_scan_obj, snpprobs_obj$snpinfo, 1:2, show_all_snps = FALSE,
         drop_hilit = 1.5, facet = "pheno")
```

```{r}
# See shinySNPPattern.R
autoplot(snp_scan_obj, snpprobs_obj$snpinfo, 1:2, 
         patterns = "hilit", drop_hilit = 1.5, facet = "pheno")
```
